Implementing a simple check to limit the maximum number of functions loaded

fabiocav opened the pull request:
Simple implementation of a function limiting check.

Resolves #170

davidebbo commented on src/WebJobs.Script/Host/ScriptHost.cs:
```diff
@@ -496,6 +496,8 @@ private static ScriptType ParseScriptType(string scriptFilePath)
 
         internal Collection<FunctionDescriptor> ReadFunctions(List<FunctionMetadata> metadatas, IEnumerable<FunctionDescriptorProvider> descriptorProviders)
         {
+            metadatas = ApplyMaximumFunctionCount(metadatas);
```
Maybe a bit simpler to ignore at the source in the `ScriptHost.ReadFunctions` loop?

fabiocav replied:
Thought about that, but wanted to keep that logic isolated and working on top of the parsed metadata.
I think you're right, though. Although this is super simple as is, we can probably have something simpler and avoid some unnecessary processing by adding this check to the loop.

davidebbo replied:
It's your call, we can stay with what you have. The whole thing is an edge case, so the unnecessary processing is hardly a concern.

davidebbo commented on src/WebJobs.Script/Host/ScriptHost.cs:
```diff
@@ -525,6 +527,27 @@ internal Collection<FunctionDescriptor> ReadFunctions(List<FunctionMetadata> met
             return functionDescriptors;
         }
 
+        private List<FunctionMetadata> ApplyMaximumFunctionCount(List<FunctionMetadata> metadatas)
```
`functions` is probably a clearer name than `metadatas`. Besides, `metadatas` is not a valid plural form for `metadata`. ;)

fabiocav replied:
Agree! `metadatas` was bugging me, but was already there, so I didn't want to change :) I'll update it throughout.

mathewc replied:
Yeah, "metadatas" was a relic from a large scale refactoring I did a long time ago. I didn't like it either :)

fabiocav commented:
@davidebbo , changed the implementation and moved the check here:
https://github.com/Azure/azure-webjobs-sdk-script/pull/177/files?diff=unified#diff-312918ec1cd07ca8bb15fe41eb8836b6R406

@mathewc in addition to the regular feedback, let me know if you have concerns about the variable name changes in the PR

mathewc commented on src/WebJobs.Script/Config/ScriptHostConfiguration.cs:
```diff
@@ -10,12 +10,15 @@ namespace Microsoft.Azure.WebJobs.Script
             RootScriptPath = Environment.CurrentDirectory;
-            RootLogPath = Path.Combine(Path.GetTempPath(), "Functions");            
+            RootLogPath = Path.Combine(Path.GetTempPath(), "Functions");
+            MaximumFunctionCount = DefaultMaximumFunctionCount;
```
nit: MaxFunctionCount. Matches the core SDK naming scheme (e.g. config.Queues.MaxDequeueCount)

mathewc commented on src/WebJobs.Script/Host/ScriptHost.cs:
```diff
@@ -403,11 +403,18 @@ private Collection<FunctionDescriptor> ReadFunctions(ScriptHostConfiguration con
+                    if (functions.Count == ScriptConfig.MaximumFunctionCount)
+                    {
+                        AddFunctionError(functionName, string.Format(CultureInfo.InvariantCulture,
+                            "Maximun number of functions ({0}) exceeded. Ignoring function.", ScriptConfig.MaximumFunctionCount));
```
spelling: Maximun

mathewc commented on src/WebJobs.Script/Host/ScriptHost.cs:
```diff
@@ -403,11 +403,18 @@ private Collection<FunctionDescriptor> ReadFunctions(ScriptHostConfiguration con
+                    {
+                        AddFunctionError(functionName, string.Format(CultureInfo.InvariantCulture,
+                            "Maximun number of functions ({0}) exceeded. Ignoring function.", ScriptConfig.MaximumFunctionCount));
+                        continue;
```
Don't we just want to `break` here? Emit a single log entry "The maximum number of functions (N) has been exceeded. Skipping the remaining (M) functions."

davidebbo replied:
Yes, it should be `break`, unless we're going to display the name of all the skipped functions (which is probably unnecessary).

fabiocav replied:
I considered a `break` but wanted to log each ignored function, since currently, we're not explicitly applying any kind of ordering and it would probably be helpful to know which functions are not running. I guess we also log which functions **are** running, so that may not be as important. I actually have a check in the test to make sure that is being logged as expected.
I'm don't have a strong opinion on this, so if you guys think the logging is irrelevant, I can change this to break out of the loop.

mathewc commented on test/WebJobs.Script.Tests/ScriptHostTests.cs:
```diff
@@ -125,5 +128,33 @@ public void ApplyFunctionsFilter()
         }
+
+        [Fact]
+        public void HostLimitsNumberOfFunctionsToMaximumcount()
```
nit: ...MaxFunctionCount

fabiocav commented:
@mathewc @davidebbo , updated based on feedback. I kept the existing logic, not breaking out of the loop, but as I mentioned, I'd be happy to break out of the loop early as I don't think we'd lose much by doing so and it would be a more optimized approach.

fabiocav commented:
Want to confirm; based on the information I got from @mathewc, we won't be moving forward with this change as it is something that will be done outside of the host. Is that correct? If so, I'll close this PR and issue #170

davidebbo commented:
Yes, that is correct. Sorry for the time you spent of this. We hadn't fully thought through it and the change in runtime would not have actually achieved the safety we were after.
